centrallix/expression/exp_functions.c

```diff
@@ -3935,146 +3936,230 @@ int exp_fn_lev_compare(pExpression tree, pParamObjects objlist, pExpression i0,
+ *
+ * Returns:
+ * 	vector table index (integer)
+ */
+int exp_fn_i_hash_char_pair(double num1, double num2)
+    {
+    int func_result = round(((num1 * num1 * num1) + (num2 * num2 * num2)) * ((num1+1)/(num2+1))) -1;
+    return func_result % EXP_VECTOR_TABLE_SIZE;	
+    }
 
-int exp_fn_i_frequency_table(double *table, char *term)
-	{
-		int i;
-		// Initialize hash table with complete character set and 0 values
-		for (i = 0; i < strlen(CHAR_SET); i++) {
-			table[i] = 0.0;
-		}
-		
-		// Iterate through term and update hash table data
-		for (i = 0; i < strlen(term); i++) {
-			// Locate index position based on where letter in term is inside the CHAR_SET
-			// Used so that CHAR_SET can be arbitrarily extended.
-			char *loc = strchr(CHAR_SET, toupper(term[i]));
-			if (loc) {
-				int index = (int)(loc - CHAR_SET);
-				table[index]++;
-			}
-		}
 
-		return 0;
+/*
+ * exp_fn_i_frequency_table
+ * This method creates a vector frequency table based on a string of characters.
+ *
+ * Parameters:
+ * 	table : integer pointer to vector frequency table
+ * 	term : the string of characters
+ *
+ * Returns:
+ * 	0 	
+ */
+int exp_fn_i_frequency_table(unsigned char *table, char *term)
+    {
+    int i;
+    // Initialize hash table with 0 values
+    for (i = 0; i < EXP_VECTOR_TABLE_SIZE; i++)
+	{
+	table[i] = 0;
 	}
 
-int exp_fn_i_relative_frequency_table(double *frequency_table)
+	int j = -1;
+    for(i = 0; i < strlen(term) + 1; i++)
 	{
-		int i;
-		double sum = 0;
-		// Compute the total character frequency
-		for (i = 0; i < strlen(CHAR_SET); i++) {
-			sum += frequency_table[i];
-		}
+	// If latter character is punctuation or whitespace, skip it
+	if (ispunct(term[i]) || isspace(term[i]))
+	    {
+	    continue;
+	    }
 
-		for (i = 0; i < strlen(CHAR_SET); i++) {
-			frequency_table[i] = frequency_table[i] / sum;
-		}
-		return 0;
-	}
+	double temp1 = 0.0;
+	double temp2 = 0.0;
 
-int exp_fn_i_tf_idf_table(double *frequency_table)
-	{
-		int i;
-		double sum = 0;
-		// Compute the total character frequency
-		for (i = 0; i < strlen(CHAR_SET); i++) {
-			sum += frequency_table[i];
-		}
+	// If previous character is null
+	if (j == -1)
+	    {
+	    temp1 = 96;
+	    }
+
+	// Else character is not null
+	else
+	    {
+	    temp1 = (int)tolower(term[j]);
+	    }
+
+	// If latter character is null
+	if (i == strlen(term))
+	    {
+	    temp2 = 96;
+	    }
+
+	// Else character is not null
+	else
+	    {
+	    temp2 = (int)tolower(term[i]);
+	    }
+
+	// Else character is not null	// If either character is a number, reassign the code point
+	if (temp1 >= 48 && temp1 <= 57)
+	    {
+	    temp1 += 75;
+	    }
+
+	if (temp2 >= 48 && temp2 <= 57)
+	    {
+	    temp2 += 75;
+	    }
+
+	// Hash the character pair into an index
+	int index = exp_fn_i_hash_char_pair(temp1, temp2);
+
+	// Increment Frequency Table value by number from 0 to 13
+	table[index] += ((unsigned char)temp1 + (unsigned char)temp2) % 13;
```

Comment:
table[] holds type unsigned char, which depending on the string could wrap around with sufficient frequency additions.  Formerly the table[] values were doubles.  I'm wondering if the table data type should be a **short** (16 bit) anyhow.

Also, it would make sense to ensure that the table increment is nonzero.  So maybe n%13 + 1 instead of just n%13?